### PR TITLE
OHRM5X-1555: Fix recruitment - candidate, vacancy related issues

### PR DESCRIPTION
**Pull request by SandeepaOHRM**
This PR Contains:-

  - OHRM5X-1555
  - OHRM5X-1556 
  - OHRM5X-1539 
  - OHRM5X-1566 
  - OHRM5X-1593
  - OHRM5X-1408

### CODE REVIEW

File: src/plugins/orangehrmRecruitmentPlugin/Api/VacancyAPI.php [hunk 1]
```diff
@@ -188,9 +189,35 @@ public function create(): EndpointResult
...
+        if (!$this->getVacancyService()->getVacancyDao()->isActiveJobVacancy($jobTitleId) instanceof JobTitle) {
+            throw $this->getBadRequestException('Please Select An Active Job Title');
+        }
+
+        $hiringManagerId = $this->getRequestParams()->getInt(
+            RequestParams::PARAM_TYPE_BODY,
+            self::PARAMETER_EMPLOYEE_ID
+        );
+        if (!$this->getVacancyService()->getVacancyDao()->isActiveHiringManger($hiringManagerId) instanceof Employee) {
+            throw $this->getBadRequestException('Employee No Longer Exists');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/VacancyAPI.php:209]: is this message showing in FE toast?

File: src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php [hunk 2]
```diff
@@ -50,4 +60,26 @@ public function preRender(Request $request): void
         $component->addProp(new Prop('history-id', Prop::TYPE_NUMBER, $historyId));
         $this->setComponent($component);
     }
+
+    public function isCapable(Request $request): bool
+    {
+        if ($request->attributes->getInt('candidateId') && $request->attributes->getInt('historyId')) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php:66]:
```suggestion
        if ($request->attributes->has('candidateId') && $request->attributes->has('historyId')) {
```

File: src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php [hunk 3]
```diff
@@ -50,4 +60,26 @@ public function preRender(Request $request): void
...
+                return false;
+            }
+            if (!$this->getUserRoleManager()->isEntityAccessible(CandidateHistory::class, $historyId)) {
+                return false;
+            }
+            if (!$this->getCandidateService()
+                    ->getCandidateDao()
+                    ->getCandidateHistoryRecordByCandidateIdAndHistoryId($candidateId, $historyId)
+                instanceof CandidateHistory) {
+                return false;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php:79]: I feel here it should be no record found

File: src/plugins/orangehrmRecruitmentPlugin/Dao/RecruitmentAttachmentDao.php [hunk 4]
```diff
@@ -266,21 +266,29 @@ public function getInterviewAttachmentListForHiringManger(int $empNumber): array
      * @return int[]
      */
     public function getInterviewAttachmentListForInterviewer(int $empNumber): array
+    {
+        $q = $this->createQueryBuilder(InterviewAttachment::class, 'attachment');
+        $q->leftJoin('attachment.interview', 'interview');
+        $q->select('attachment.id');
+        $q->andWhere('interview IN (:ids)');
+        $q->andWhere($q->expr()->in('interview', ':ids'));
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/RecruitmentAttachmentDao.php:274]:
```suggestion
        $q->andWhere($q->expr()->in('interview', ':ids'));
```
:open_mouth:

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php [hunk 5]
```diff
@@ -141,12 +138,7 @@ public function deleteVacancies(array $toBeDeletedVacancyIds): bool
         $qr->delete()
             ->andWhere('v.id IN (:ids)')
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:139]:
```suggestion
            ->andWhere($q->expr()->in('v.id', ':ids'))
```

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php [hunk 6]
```diff
@@ -211,4 +203,27 @@ public function getPublishedVacancyList(): array
         $qb->addOrderBy('vacancy.updatedTime', ListSorter::DESCENDING);
         return $qb->getQuery()->execute();
     }
+
+    /**
+     * @param int $jobTitleId
+     * @return JobTitle|null
+     */
+    public function isActiveJobVacancy(int $jobTitleId): ?JobTitle
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:211]:
```suggestion
    public function isActiveJobTitle(int $jobTitleId): ?JobTitle
```
return type also wrong with the method name. if you want to get job title also (not a boolean) use `JobTitleService`

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php [hunk 7]
```diff
@@ -211,4 +203,27 @@ public function getPublishedVacancyList(): array
...
+    {
+        return $this->getRepository(JobTitle::class)->findOneBy(['id' => $jobTitleId, 'isDeleted' => false]);
+    }
+
+    /**
+     * @param int $hiringManagerId
+     * @return Employee|null
+     */
+    public function isActiveHiringManger(int $hiringManagerId): ?Employee
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:220]: here also return type should be boolean according to the method name
```php
        return $this->getRepository(Employee::class)->findOneBy(
            [
                'empNumber' => $hiringManagerId,
                'employeeTerminationRecord' => null
            ]
        ) instanceof Employee;
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmRecruitmentPlugin/Api/VacancyAPI.php (hunk 1 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/VacancyAPI.php:209]: no

File: src/plugins/orangehrmRecruitmentPlugin/Dao/RecruitmentAttachmentDao.php (hunk 4 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/RecruitmentAttachmentDao.php:274]: oops forgot to remove 273

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php (hunk 5 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:139]: forgot to remove this also

File: src/client/src/orangehrmRecruitmentPlugin/pages/AddJobVacancy.vue [hunk 8]
```diff
@@ -212,7 +212,7 @@ export default {
         name: this.vacancy.name,
         jobTitleId: this.vacancy.jobTitle.id,
         employeeId: this.vacancy.hiringManager.id,
-        numOfPositions: this.vacancy.numOfPositions,
+        numOfPositions: this.vacancy.numOfPositions || null,
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/client/src/orangehrmRecruitmentPlugin/pages/AddJobVacancy.vue:215]: this is not working. numOfPositions getIntOrNull and its type validation doesn't allow empty string to exclude

File: src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue [hunk 9]
```diff
@@ -458,18 +459,20 @@ export default {
         this.currentName = data.name;
         this.vacancy.name = data.name;
         this.vacancy.description = data.description;
-        this.vacancy.numOfPositions = data.numOfPositions;
+        this.vacancy.numOfPositions = data.numOfPositions || '';
         this.vacancy.status = data.status === 1 ? true : false;
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue:463]: done

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php (hunk 6 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:211]: fixed

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php (hunk 7 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:220]: fixed

File: src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue:463]: fixed

File: src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php (hunk 3 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php:79]: when manually changing the url with history id, the 4x behaviour is it shows required credentials message, if the particular candidate does not have a record with the historyId. if we throw record not found, it keeps loading.

File: src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php (hunk 2 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Controller/WorkflowActionHistoryController.php:66]: done

File: src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue [hunk 10]
```diff
@@ -367,7 +368,16 @@ export default {
...
-        hiringManager: [required],
+        hiringManager: [
+          required,
+          v => {
+            if (v?.isPastEmployee) {
+              return this.$t('general.invalid');
+            } else {
+              return true;
+            }
+          },
```
Comment (RajithaKumara) [inline src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue:379]:
```suggestion
          v => (v?.isPastEmployee ? this.$t('general.invalid'): true;)
```
:thinking:

File: src/plugins/orangehrmRecruitmentPlugin/Api/CandidateHistoryAPI.php [hunk 11]
```diff
@@ -194,6 +194,7 @@ public function getValidationRuleForUpdate(): ParamRuleCollection
                     self::PARAMETER_NOTE,
                     new Rule(Rules::STRING_TYPE)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/CandidateHistoryAPI.php:195]: don't we have FE text length validation? please add length validation

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php [hunk 12]
```diff
@@ -211,4 +203,32 @@ public function getPublishedVacancyList(): array
...
+     * @param int $jobTitleId
+     * @return bool
+     */
+    public function isActiveJobVacancy(int $jobTitleId): bool
+    {
+        return $this->getRepository(JobTitle::class)->findOneBy(
+            [
+                    'id' => $jobTitleId,
+                    'isDeleted' => false
+                ]
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:217]: just double check formatting here

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php [hunk 13]
```diff
@@ -211,4 +203,32 @@ public function getPublishedVacancyList(): array
...
+     * @param int $hiringManagerId
+     * @return bool
+     */
+    public function isActiveHiringManger(int $hiringManagerId): bool
+    {
+        return $this->getRepository(Employee::class)->findOneBy(
+            [
+                    'empNumber' => $hiringManagerId,
+                    'employeeTerminationRecord' => null
+                ]
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:231]: format the code

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue (hunk 10 above)
Comment (SandeepaOHRM) [reply src/client/src/orangehrmRecruitmentPlugin/pages/EditJobVacancy.vue:379]: done

File: src/plugins/orangehrmRecruitmentPlugin/Api/CandidateHistoryAPI.php (hunk 11 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/CandidateHistoryAPI.php:195]: oops

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php (hunk 12 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:217]: done

File: src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php (hunk 13 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/VacancyDao.php:231]: done

File: src/plugins/orangehrmRecruitmentPlugin/Api/CandidateHistoryAPI.php (hunk 11 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/CandidateHistoryAPI.php]: done

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**